### ISSUE TITLE
Large delay when loading huge webvtt subtitle file

### ISSUE THREAD

**Issue by epigramx** [clsid2/mpc-hc#420]
Example video: https://www.youtube.com/watch?v=G5B1mIfQuo4

Receive it locally with youtube-dl: https://github.com/ytdl-org/youtube-dl

**Comment by epigramx** [comment]
~~Re-"containing" it with ffmpeg to a new mkv (or mp4) with a copy of the streams appears to alleviate the issue. So it's not the streams that cause it.~~

**Comment by epigramx** [comment]
False alarm. It's the SUBTITLES that cause it (you can get them with youtube-dl --write-auto-sub --sub-lang en).

**Comment by clsid2** [comment]
Upload the subtitle file somewhere. That should be enough to reproduce.

**Comment by epigramx** [comment]
Yes. It's definitely because of the WebVTT subtitles. It goes away with converting them to subrip.
[YMS - Kimba the White Lion-G5B1mIfQuo4.en.zip](https://github.com/clsid2/mpc-hc/files/4696678/YMS.-.Kimba.the.White.Lion-G5B1mIfQuo4.en.zip)

**Comment by clsid2** [comment]
@adipose
Maybe (most of) the regex_replace processing could be done just before the actual rendering of a subtitle line, instead of doing it all when loading the file?

**Comment by adipose** [comment]
Yes, theoretically, but there are some concerns.

1. CSimpleTextSubtitle::Open() calls OpenFuncts[] which is generalized for several subtitle types, meaning any function that implements it is supposed to return a buffer of valid subtitles.
2. All subtitles are validated when they are loaded, and rejected if the file doesn't parse properly.
3. Other code is run subsequently in ::Open() (SSA styles file) which needs to append to the original subtitles.

Basically what it's doing is converting to SSA, after cleaning up VTT to look like SRT.  To parse this on the fly would require some sort of intelligent look-ahead, as you need to know which lines are needed for the current timestamps.

However, if SRT is faster, maybe the algorithm is bad for VTT.  SRT uses a similar function to convert to SSA.

**Comment by adipose** [comment]
Test build:  https://mega.nz/file/NBABFAgR#3H6PkAtoTA69bsN49q2_YkBKhjgd1p3OC3xQpuMZlws

The big delay comes from the OpenRealText function.  I put WEBVTT before that in the sequence, and now it's much faster.

I didn't look closely, but apparently OpenRealText thinks the WEBVTT is worth trying to parse, and it takes a very long time to do it.  Some of the functs drop out quickly based on a header (as I designed for WEBVTT), but others aren't able to fail so easily.

Perhaps a blacklist should be implemented (such as: file extension is a known type=vtt, do not try to load x, y or z), because as of now, it's trying all types sequentially, and surely some are unneeded.

Another improvement would be identifying the preferred function, trying that first, and then (if desired) try some of the others.

https://github.com/clsid2/mpc-hc/pull/480

**Comment by clsid2** [comment]
Yes, that is a good idea. Both can be combined in a single function that determines which parsers should be tried and in which order, based on the file extension.

For example for .vtt:
First try WEBVTT. Second SRT as a fallback. Others are not relevant.

The uncommon generic parsers only need to be used in case of .srt .txt and unknown extensions.
